File: src/find_missing_elements_3731.rs
// https://leetcode.com/problems/find-missing-elements

use itertools::Itertools;

pub fn find_missing_elements(mut nums: Vec<i32>) -> Vec<i32> {
    nums.sort_unstable();
    nums.into_iter()
        .tuple_windows()
        .flat_map(|(a, b)| (a + 1)..b)
        .collect()
}

/// Alternative: mark presence in a `u64` bitset, then scan the full `[min, max]`
/// range collecting values whose bit is unset.
pub fn find_missing_elements_bitset(nums: Vec<i32>) -> Vec<i32> {
    let (&min, &max) = (
        nums.iter().min().unwrap(),
        nums.iter().max().unwrap(),
    );
    let span = (max - min + 1) as usize;
    let mut bits = vec![0u64; span.div_ceil(64)];
    for &v in &nums {
        let i = (v - min) as usize;
        bits[i / 64] |= 1u64 << (i % 64);
    }
    (min..=max)
        .filter(|&v| {
            let i = (v - min) as usize;
            bits[i / 64] & (1u64 << (i % 64)) == 0
        })
        .collect()
}

/// Alternative: mark presence in a boolean array, then scan the full `[min, max]`
/// range collecting values that are still `false`.
pub fn find_missing_elements_bool(nums: Vec<i32>) -> Vec<i32> {
    let (&min, &max) = (
        nums.iter().min().unwrap(),
        nums.iter().max().unwrap(),
    );
    let span = (max - min + 1) as usize;
    let mut present = vec![false; span];
    for &v in &nums {
        present[(v - min) as usize] = true;
    }
    (min..=max)
        .filter(|&v| !present[(v - min) as usize])
        .collect()
}

/// Alternative: sort, then imperatively push every value strictly between
/// consecutive elements.
pub fn find_missing_elements_loop(mut nums: Vec<i32>) -> Vec<i32> {
    nums.sort_unstable();
    let mut ans = Vec::new();
    for i in 0..nums.len() - 1 {
        for x in nums[i] + 1..nums[i + 1] {
            ans.push(x);
        }
    }
    ans
}

/// Alternative: pack presence into a single `u128` bitmask, then walk from the
/// lowest set bit to the highest, emitting positions whose bit is unset. Only
/// valid when all values fit in `0..128` (true for this problem: `1 <= v <= 100`).
pub fn find_missing_elements_u128(nums: Vec<i32>) -> Vec<i32> {
    let b = nums.into_iter().fold(0u128, |f, n| f | 1 << n);

    std::iter::successors(Some(b), |n| Some(n >> 1))
        .skip(b.trailing_zeros() as _)
        .zip(b.trailing_zeros() as i32..127 - b.leading_zeros() as i32)
        .filter_map(|(b, m)| (b & 1 == 0).then_some(m))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::{
        find_missing_elements, find_missing_elements_bitset,
        find_missing_elements_bool, find_missing_elements_loop,
        find_missing_elements_u128,
    };

    #[test]
    fn official1() {
        assert_eq!(vec![3], find_missing_elements(vec![1, 4, 2, 5]));
    }

    #[test]
    fn official2() {
        assert_eq!(vec![] as Vec<i32>, find_missing_elements(vec![7, 8, 6, 9]));
    }

    #[test]
    fn official3() {
        assert_eq!(vec![2, 3, 4], find_missing_elements(vec![5, 1]));
    }

    #[test]
    fn bitset_matches_official() {
        assert_eq!(vec![3], find_missing_elements_bitset(vec![1, 4, 2, 5]));
        assert_eq!(vec![] as Vec<i32>, find_missing_elements_bitset(vec![7, 8, 6, 9]));
        assert_eq!(vec![2, 3, 4], find_missing_elements_bitset(vec![5, 1]));
    }

    #[test]
    fn bool_matches_official() {
        assert_eq!(vec![3], find_missing_elements_bool(vec![1, 4, 2, 5]));
        assert_eq!(vec![] as Vec<i32>, find_missing_elements_bool(vec![7, 8, 6, 9]));
        assert_eq!(vec![2, 3, 4], find_missing_elements_bool(vec![5, 1]));
    }

    #[test]
    fn u128_matches_official() {
        assert_eq!(vec![3], find_missing_elements_u128(vec![1, 4, 2, 5]));
        assert_eq!(vec![] as Vec<i32>, find_missing_elements_u128(vec![7, 8, 6, 9]));
        assert_eq!(vec![2, 3, 4], find_missing_elements_u128(vec![5, 1]));
    }

    #[test]
    fn loop_matches_official() {
        assert_eq!(vec![3], find_missing_elements_loop(vec![1, 4, 2, 5]));
        assert_eq!(vec![] as Vec<i32>, find_missing_elements_loop(vec![7, 8, 6, 9]));
        assert_eq!(vec![2, 3, 4], find_missing_elements_loop(vec![5, 1]));
    }
}

File: cpp/src/find_missing_elements_3731.cpp
// https://leetcode.com/problems/find-missing-elements

#include "test.hpp"

#include <algorithm>
#include <cstdint>
#include <ranges>
#include <unordered_set>
#include <vector>

std::vector<int> findMissingElements(std::vector<int> nums) {
    std::ranges::sort(nums);
    std::vector<int> result;
    // `zip` with a dropped-by-one copy is the pairwise view libc++ lacks; it also
    // sidesteps the `nums.size() - 1` underflow on an empty input.
    for (const auto [lo, hi] : std::views::zip(nums, nums | std::views::drop(1))) {
        result.append_range(std::views::iota(lo + 1, hi));
    }
    return result;
}

// Alternative: mark presence in a `std::vector<bool>`, then scan the full `[min, max]`
// range collecting values that are still `false`.
std::vector<int> findMissingElementsBool(std::vector<int> nums) {
    const auto [min_val, max_val] = std::ranges::minmax(nums);
    const auto span = static_cast<std::size_t>(max_val - min_val + 1);

    std::vector<bool> present(span, false);
    for (int v : nums) {
        present[static_cast<std::size_t>(v - min_val)] = true;
    }

    std::vector<int> result;
    for (int v = min_val; v <= max_val; ++v) {
        if (!present[static_cast<std::size_t>(v - min_val)]) {
            result.push_back(v);
        }
    }
    return result;
}

// Alternative: collect into an `std::unordered_set`, then scan the full open range
// `(min, max)` collecting values absent from the set.
std::vector<int> findMissingElementsHashSet(const std::vector<int>& nums) {
    std::unordered_set<int> st(nums.begin(), nums.end());
    const auto [min_val, max_val] = std::ranges::minmax(nums);

    std::vector<int> result;
    for (int x = min_val + 1; x < max_val; ++x) {
        if (st.find(x) == st.end()) {
            result.push_back(x);
        }
    }
    return result;
}

// Alternative: mark presence in a bitset, then scan the full `[min, max]`
// range collecting values whose bit is unset.
std::vector<int> findMissingElementsBitset(const std::vector<int>& nums) {
    const auto [min_val, max_val] = std::ranges::minmax(nums);
    const auto span = static_cast<std::size_t>(max_val - min_val + 1);
    const auto num_words = (span + 63) / 64;
    std::vector<std::uint64_t> bits(num_words, 0);

    for (int v : nums) {
        const auto i = static_cast<std::size_t>(v - min_val);
        bits[i / 64] |= 1ULL << (i % 64);
    }

    std::vector<int> result;
    for (int v = min_val; v <= max_val; ++v) {
        const auto i = static_cast<std::size_t>(v - min_val);
        if ((bits[i / 64] & (1ULL << (i % 64))) == 0) {
            result.push_back(v);
        }
    }
    return result;
}

TEST(official1) {
    check_eq({3}, findMissingElements({1, 4, 2, 5}));
}

TEST(official2) {
    check_eq(std::vector<int>{}, findMissingElements({7, 8, 6, 9}));
}

TEST(official3) {
    check_eq({2, 3, 4}, findMissingElements({5, 1}));
}

TEST(bool_matches_official1) {
    check_eq({3}, findMissingElementsBool({1, 4, 2, 5}));
}

TEST(bool_matches_official2) {
    check_eq(std::vector<int>{}, findMissingElementsBool({7, 8, 6, 9}));
}

TEST(bool_matches_official3) {
    check_eq({2, 3, 4}, findMissingElementsBool({5, 1}));
}

TEST(hashset_matches_official1) {
    check_eq({3}, findMissingElementsHashSet({1, 4, 2, 5}));
}

TEST(hashset_matches_official2) {
    check_eq(std::vector<int>{}, findMissingElementsHashSet({7, 8, 6, 9}));
}

TEST(hashset_matches_official3) {
    check_eq({2, 3, 4}, findMissingElementsHashSet({5, 1}));
}

TEST(bitset_matches_official1) {
    check_eq({3}, findMissingElementsBitset({1, 4, 2, 5}));
}

TEST(bitset_matches_official2) {
    check_eq(std::vector<int>{}, findMissingElementsBitset({7, 8, 6, 9}));
}

TEST(bitset_matches_official3) {
    check_eq({2, 3, 4}, findMissingElementsBitset({5, 1}));
}
